### ISSUE TITLE
Pop queue not working as expected

### ISSUE THREAD

**Issue by alchitry** [sparkfun/Qwiic_Button#1]
### Subject of the issue
I'm trying to get the Qwiic button working with the RedBoard Turbo and have setup all the libraries and am able to talk to the button.

Stuff like button.isPressed() works exactly as expected. However, the queue stuff isn't quite working properly.

I thought I'd use the pressed queue to make sure I got an accurate count of how many times the button was pressed. I'm using button.isPressedQueueEmpty() to check if the queue is empty before calling button.popPressedQueue() to remove a press. However, popPressedQueue() doesn't seem to remove an element from the queue as expected.

After clicking the button once, isPressedQueueEmpty() returns false regardless of how many times popPressedQueue() is called. Also isPressedQueueFull() always returns true once I press the button 15 times.

I've checked the firmware of the button and it matches the latest in the GitHub (258, 0x01 major, 0x02 minor)

I've also tried writing 0x01 to pressedQueueStatus register directly and reading it back. It reads 0 so the popRequest bit is either never set or cleared when the request is supposedly processed.

### Your workbench
* What development board or microcontroller are you using?
RedBoard Turbo

* What version of hardware or breakout board are you using? 
258 (current firmware)

* How is the breakout board wired to your microcontroller?
Qwiic

* How is everything being powered?
5V regulator to the RedBoard Turbo, button is powered over Qwiic

### Steps to reproduce
Connect button to RedBoard Turbo (or likely any Qwiic board).

Here's the code I used.

```
#include <SparkFun_Qwiic_Button.h>

const int BLUE_LED = 13;
const int RX_LED = PIN_LED_RXL; // RX LED on pin 25, we use the predefined PIN_LED_RXL to make sure
const int TX_LED = PIN_LED_TXL; // TX LED on pin 26, we use the predefined PIN_LED_TXL to make sure

QwiicButton hourUpButton;
QwiicButton hourDownButton;
QwiicButton minuteUpButton;
QwiicButton minuteDownButton;

void setup() {
  pinMode(BLUE_LED, OUTPUT);
  pinMode(RX_LED, OUTPUT);
  pinMode(TX_LED, OUTPUT);
  digitalWrite(RX_LED, LOW); // TX LED on

  SerialUSB.begin(9600);
  Wire.begin(); //Join I2C bus

  if (minuteDownButton.begin(0x67) == false)
    errorLoop();

  digitalWrite(RX_LED, HIGH); // TX LED off

  while (!SerialUSB);
  digitalWrite(BLUE_LED, HIGH); // Blue LED on
  SerialUSB.println("Starting");

}

void loop() {
  if (minuteDownButton.isPressedQueueEmpty() == false) {
    SerialUSB.println("Minute Down");
    minuteDownButton.popPressedQueue();
  }
  delay(100);
}

void errorLoop() {
  while (1) {
    digitalWrite(RX_LED, LOW); // RX LED on
    delay(333);
    digitalWrite(RX_LED, HIGH); // RX LED off
    digitalWrite(TX_LED, LOW); // TX LED on
    delay(333);
    digitalWrite(TX_LED, HIGH); // TX LED off
    digitalWrite(BLUE_LED, HIGH); // Blue LED on
    delay(333);
    digitalWrite(BLUE_LED, LOW); // Blue LED off
  }
}
```

### Expected behavior
Calling button.popPressedQueue() should remove a press from the queue.

The code should print "Minute Down" once per button press.

### Actual behavior
The queue isn't emptied.

The code prints "Minute Down" continuously after a single press.

**Comment by wickedbeernut** [comment]
This bug was opened six months ago.  Can you please help me understand why it hasn't been fixed?  I just bought eight QWIIC buttons and switches and they're all unusable due to this bug.

The way in which the queueStatusBitField is defined,
```
typedef union {
  struct {
    bool popRequest : 1; //This is bit 0. User mutable, user sets to 1 to pop from queue, we pop from queue and set the bit back to zero.
    bool isEmpty : 1; //user immutable, returns 1 or 0 depending on whether or not the queue is empty
    bool isFull : 1; //user immutable, returns 1 or 0 depending on whether or not the queue is full
    bool: 5;
  };
  uint8_t byteWrapped;
} queueStatusBitField;
```
is inconsistent with the way in which it's being initialized as part of the protectionMap (pressedQueueStatus and clickedQueueStatus),

```
memoryMap protectionMap = {
  0x00,       //id
  0x00,       //firmwareMinor
  0x00,       //firmwareMajor
  {1, 1, 1},  //buttonStatus {isPressed, hasBeenClicked, eventAvailable}
  {1, 1},     //interruptConfig {pressedEnable, clickedEnable}
  0xFFFF,     //buttonDebounceTime
  {0, 0, 1},  //pressedQueueStatus {isFull, isEmpty, popRequest}
  0x00000000, //pressedQueueFront
  0x00000000, //pressedQueueBack
  {0, 0, 1},  //clickedQueueStatus {isFull, isEmpty, popRequest}
  0x00000000, //clickedQueueFront
  0x00000000, //clickedQueueBack
  0xFF,       //ledBrightness
  0xFF,       //ledPulseGranularity
  0xFFFF,     //ledPulseCycleTime
  0xFFFF,     //ledPulseOffTime
  0xFF,       //i2cAddress
};
```

consequently, the popRequest bit is read-only rather than read/write.

**Comment by englandsaurus** [comment]
@wickedbeernut  Are you using a SAMD microcontroller?

**Comment by wickedbeernut** [comment]
> @wickedbeernut Are you using a SAMD microcontroller?

You got me there!  SparkFun SAMD51 Thing Plus.  

I'll withdraw my comment with respect to the byte-alignment of the memoryMap.  However, I stand by my comment with respect to the way in which the queueStatusBitField is defined and initialized as part of the registerMap and protectionMap (pressedQueueStatus and clickedQueueStatus).

I compiled and uploaded the Qwiic_Button firmware to an Uno.  I added the following Serial.print statements to setup(),

```
  Serial.println();
  Serial.println("memoryMap");
  Serial.println("  Register: registerMap protectionMap");
  for (int x = 0; x < sizeof(memoryMap); x++) {
      Serial.print("    0x");
      Serial.print(x, HEX);
      Serial.print(": ");
      Serial.print(*(registerPointer + x), BIN);
      Serial.print(" ");
      Serial.println(*(protectionPointer + x), BIN);
  }

   bool popRequest;
   bool isEmpty;
   bool isFull;

   popRequest = protectionMap.pressedQueueStatus.popRequest;
   isEmpty = protectionMap.pressedQueueStatus.isEmpty;
   isFull = protectionMap.pressedQueueStatus.isFull;
   
   Serial.println();
   Serial.println("Protection Map");
   Serial.print("  pressedQueueStatus: ");
   Serial.println(*(protectionPointer + 7), BIN);
   Serial.print("    protectionMap.popRequest: ");
   Serial.println(popRequest);
   Serial.print("    protectionMap.isEmpty: ");
   Serial.println(isEmpty);
   Serial.print("    protectionMap.isFull: ");
   Serial.println(isFull);
   Serial.println();

   popRequest = registerMap.pressedQueueStatus.popRequest;
   isEmpty = registerMap.pressedQueueStatus.isEmpty;
   isFull = registerMap.pressedQueueStatus.isFull;

   Serial.println("Register Map");
   Serial.print("  pressedQueueStatus: ");
   Serial.println(*(registerPointer + 7), BIN);
   Serial.print("    registerMap.popRequest: ");
   Serial.println(popRequest);
   Serial.print("    registerMap.isEmpty: ");
   Serial.println(isEmpty);
   Serial.print("    registerMap.isFull: ");
   Serial.println(isFull);
   Serial.println();
```

which yielded,

```
Qwiic Button
Address: 0x6F
Device ID: 0x5D

memoryMap
  Register: registerMap protectionMap
    0x0: 1011101 0
    0x1: 10 0
    0x2: 1 0
    0x3: 0 111
    0x4: 11 11
    0x5: 1010 11111111
    0x6: 0 11111111
    0x7: 10 100
    0x8: 0 0
    0x9: 0 0
    0xA: 0 0
    0xB: 0 0
    0xC: 0 0
    0xD: 0 0
    0xE: 0 0
    0xF: 0 0
    0x10: 10 100
    0x11: 0 0
    0x12: 0 0
    0x13: 0 0
    0x14: 0 0
    0x15: 0 0
    0x16: 0 0
    0x17: 0 0
    0x18: 0 0
    0x19: 0 11111111
    0x1A: 1 11111111
    0x1B: 0 11111111
    0x1C: 0 11111111
    0x1D: 0 11111111
    0x1E: 0 11111111
    0x1F: 1101111 11111111

Protection Map
  pressedQueueStatus: 100
    protectionMap.popRequest: 0
    protectionMap.isEmpty: 0
    protectionMap.isFull: 1

Register Map
  pressedQueueStatus: 10
    registerMap.popRequest: 0
    registerMap.isEmpty: 1
    registerMap.isFull: 0
```

As you can see, memoryMap is byte-aligned properly, however, the popRequest and isFull bits of the protectionMap pressedQueueStatus are read-only and read/write which is just the opposite of what they should be.  The protectionMap pressedQueueStatus (and buttonStatus, interruptConfig and clickedQueueStatus) should have been initialized as,

```
//Variables used in the I2C interrupt.ino file so we use volatile
volatile memoryMap registerMap {
  DEVICE_ID,           //id
  FIRMWARE_MINOR,      //firmwareMinor
  FIRMWARE_MAJOR,      //firmwareMajor
//{0, 0, 0},           //buttonStatus {isPressed, hasBeenClicked, eventAvailable}  INCORRECT
  {0, 0, 0},           //buttonStatus {eventAvailable, hasBeenClicked, isPressed}  CORRECT
//{1, 1},              //interruptConfig {pressedEnable, clickedEnable}  INCORRECT
  {1, 1},              //interruptConfig {clickedEnable, pressedEnable}  CORRECT
  0x000A,              //buttonDebounceTime
//{0, 1, 0},           //pressedQueueStatus {isFull, isEmpty, popRequest}  INCORRECT
  {0, 1, 0},           //pressedQueueStatus {popRequest, isEmpty, isFull}  CORRECT
  0x00000000,          //pressedQueueFront
  0x00000000,          //pressedQueueBack
//{0, 1, 0},           //clickedQueueStatus {isFull, isEmpty, popRequest}  INCORRECT
  {0, 1, 0},           //clickedQueueStatus {popRequest, isEmpty, isFull}  CORRECT
  0x00000000,          //clickedQueueFront
  0x00000000,          //clickedQueueBack
  0x00,                //ledBrightness
  0x01,                //ledPulseGranularity
  0x0000,              //ledPulseCycleTime
  0x0000,              //ledPulseOffTime
  DEFAULT_I2C_ADDRESS, //i2cAddress
};

//This defines which of the registers are read-only (0) vs read-write (1)
memoryMap protectionMap = {
  0x00,       //id
  0x00,       //firmwareMinor
  0x00,       //firmwareMajor
//{1, 1, 1},  //buttonStatus {isPressed, hasBeenClicked, eventAvailable}  INCORRECT
  {1, 1, 1},  //buttonStatus {eventAvailable, hasBeenClicked, isPressed}  CORRECT
//{1, 1},     //interruptConfig {pressedEnable, clickedEnable}  INCORRECT
  {1, 1},     //interruptConfig {clickedEnable, pressedEnable}  CORRECT
  0xFFFF,     //buttonDebounceTime
//{0, 0, 1},  //pressedQueueStatus {isFull, isEmpty, popRequest}  INCORRECT
  {1, 0, 0},  //pressedQueueStatus {popRequest, isEmpty, isFull}  CORRECT
  0x00000000, //pressedQueueFront
  0x00000000, //pressedQueueBack
//{0, 0, 1},  //clickedQueueStatus {isFull, isEmpty, popRequest}  INCORRECT
  {1, 0, 0},  //clickedQueueStatus {popRequest, isEmpty, isFull}  CORRECT
  0x00000000, //clickedQueueFront
  0x00000000, //clickedQueueBack
  0xFF,       //ledBrightness
  0xFF,       //ledPulseGranularity
  0xFFFF,     //ledPulseCycleTime
  0xFFFF,     //ledPulseOffTime
  0xFF,       //i2cAddress

};
```

**Note**:  All of the bit assignments other than pressedQueueStatus are symmetric so, consequently, there were no other negative sideeffects.  Only popRequest and isFull were reversed with respect to their read/write protection.

So with respect to my original question ... is there any particular reason why this hasn't been fixed (regardless of whether this is the issue)?  It's extremely disappointing that SparkFun continues to sell QWIIC buttons and switches with the queue functionality broken with not even a disclaimer in the product description.

**Comment by wickedbeernut** [comment]
I made the change to the protectionMap pressedQueueStatus and clickedQueueStatus as described above, namely,

```
memoryMap protectionMap = {
  ...
  {1, 0, 0},  //pressedQueueStatus {popRequest, isEmpty, isFull} 
  ...
  {1, 0, 0},  //clickedQueueStatus {popRequest, isEmpty, isFull} 
  ...
};
```
I tested the fix with a SparkFun SAMD51 Thing Plus serving as the I2C master / controller running the SparkFun_Qwiic_Button_Arduino_Library,

```
#include <SparkFun_Qwiic_Button.h>
QwiicButton button;

#define DEBOUNCE_TIME      500
#define POLLING_INTERVAL 10000

void setup() {
   Serial.begin(115200);
   delay(2000);
   Wire.begin();
   button.begin();
}

void loop() {
   int n = 0;
   unsigned long current_time = millis();
   while (button.isPressedQueueEmpty() == false) {
      Serial.print(++n);
      Serial.print(": timeSinceFirstPress: ");
      unsigned long timeSinceFirstPress = button.popPressedQueue();
      Serial.print(timeSinceFirstPress);
      if ((timeSinceFirstPress < DEBOUNCE_TIME) || (timeSinceFirstPress > POLLING_INTERVAL))
         Serial.print(" Invalid");
      Serial.println();
   }
   delay(POLLING_INTERVAL);
}
```

and an Uno serving as the QWIIC button / switch ATtiny84 running the sparkfun / Qwiic_Button firmware with the two-line fix.  Here are the results,

```
1: timeSinceFirstPress: 6227

1: timeSinceFirstPress: 9837
2: timeSinceFirstPress: 9292

1: timeSinceFirstPress: 2014
2: timeSinceFirstPress: 1503
3: timeSinceFirstPress: 891

1: timeSinceFirstPress: 120576 Invalid
1: timeSinceFirstPress: 150544 Invalid
1: timeSinceFirstPress: 12 Invalid
```

This is one button / switch press, followed by two presses, followed by three presses ... each in a ten second interval.

I had a problem with isPressedQueueEmpty() returning false with the queue empty.  Consequently, popPressedQueue() returned "garbage".  Conceivably, there could've been "noise" on the I2C bus.  I attempted to filter out these values (tagged with "Invalid").

I also had a problem maintaining the debounce time between the Thing Plus and Uno, so I changed the hardcoded default at the time the debouce time is read from the EEPROM,

```
  EEPROM.get(LOCATION_BUTTON_DEBOUNCE_TIME, map->buttonDebounceTime);
  if (map->buttonDebounceTime == 0xFFFF)
  {
// map->buttonDebounceTime = 10; //Default to 10ms
    map->buttonDebounceTime = 500; //Default to 500ms
    EEPROM.put(LOCATION_BUTTON_DEBOUNCE_TIME, map->buttonDebounceTime);
  }
```
I need to investigate where the debounce time is actually **set** in the EEPROM.

Long story short, if the Uno and ATtiny84 are functionally equivalent in this regard, the two-line fix seems to resolve the issue.

@englandsaurus, would you please confirm this fix and let me know how I go about getting the firmware on my eight QWIIC buttons and switches upgraded without spending $40 on the Tiny AVR programmer (which is currently backordered), IC test clip - SOIC 8-pin and related materials.  Can I mail the buttons and switches back to SparkFun to be upgraded?

![image](https://user-images.githubusercontent.com/47407361/109351208-1de44b80-7836-11eb-9c3e-3df22d6f3880.png)

**Comment by Noam3kCH** [comment]
I'm also waiting for a solution here

**Comment by makin-stuff** [comment]
@wickedbeernut @Noam3kCH I'm working on updating the firmware now. Will update and close this issue when I've got this bug fixed.

**Comment by makin-stuff** [comment]
The firmware has been updated and will be present on future Qwiic Button batches. Thanks for your help @wickedbeernut. I've reached out to @Funda in tech support to see if he can issue you some new buttons. He should be reaching out shortly. Thanks

**Comment by Funda** [comment]
@wickedbeernut Go ahead and fill out a return ticket here: https://www.sparkfun.com/returns - We will get you taken care of after the ticket has been submitted.

Cheers for now.

**Comment by wickedbeernut** [comment]
> @wickedbeernut Go ahead and fill out a return ticket here: https://www.sparkfun.com/returns - We will get you taken care of after the ticket has been submitted.
> 
> Cheers for now.

@Funda RMA submitted.  Thank you!

**Comment by wickedbeernut** [comment]
> The firmware has been updated and will be present on future Qwiic Button batches. Thanks for your help @wickedbeernut. I've reached out to @Funda in tech support to see if he can issue you some new buttons. He should be reaching out shortly. Thanks

I'd ordered,

(5) Qwiic Switch (SPX-15586)
(2) SparkFun Qwiic Button Breakout (BOB-15931)
(1) Qwiic Arcade - Red (SPX-15591)

Support (@Funda?) is offering to refund my money for the two SparkFun Qwiic button breakout boards (BOB-15931).  Apparently, I'm stuck with the five Qwiic switches (SPX-15586) and arcade button (SPX-15591) because they're "SparkX" (whatever that means).  They all have the same embedded firmware bug.

I don't need a refund.  I need buttons and switches with the new firmware (v01.03).  I don't care if I have to pay for them yet again.  How do I go about ordering Qwiic buttons and switches with the new firmware?

Please help @makin-stuff!

**Comment by wickedbeernut** [comment]
> The firmware has been updated and will be present on future Qwiic Button batches. Thanks for your help @wickedbeernut. I've reached out to @Funda in tech support to see if he can issue you some new buttons. He should be reaching out shortly. Thanks

@makin-stuff,

@Funda is refusing to issue me buttons with firmware version 1.3.  I recently purchased,

	(2) SparkFun Qwiic Button Breakout (BOB-15931)
	(5) Qwiic Switch (SPX-15586)
	(1) Qwiic Arcade - Red (SPX-15591)

Can you mail me buttons (BOB-15931) and/or switches (SPX-15586) with firmware version 1.3?  I can send you a Paypal payment?  $10 a piece plus shipping?

**Comment by makin-stuff** [comment]
Hi there @wickedbeernut. 
Thanks for all the work you did squashing the bug in the qwiic button firmware. 
Please continue to work with @Funda to process your return.
Please do not comment about this return on this Github issue anymore.
